### PR TITLE
Make _wait_all_workers() support being called for multiple times

### PR DESCRIPTION
**Pull request by xush6528**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32624 Make _wait_all_workers() support being called for multiple times**

We need this PR to resolve the issue mentioned in https://github.com/pytorch/pytorch/issues/31325#issuecomment-574918917.

The solution is for each `_wait_all_workers()` call, there is a sequence ID added, to identify different calls.

Differential Revision: [D5739520](https://our.internmc.facebook.com/intern/diff/D5739520/)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit a0544c66:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=a0544c66dff504d000e998b43bbea2daeb6b090b).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4342456) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90906284) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90906284))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
 
  git checkout -b <new-branch-name> 
 
HEAD is now at a0544c66df Make _wait_all_workers() support being called for multiple times 
+ git reset --hard a0544c66dff504d000e998b43bbea2daeb6b090b 
HEAD is now at a0544c66df Make _wait_all_workers() support being called for multiple times 
+ git merge --allow-unrelated-histories --no-edit --no-ff bd20274e8f667e7f59f1a592b8d905f89281ee7f 
Auto-merging torch/testing/_internal/distributed/rpc/rpc_test.py 
Auto-merging torch/distributed/rpc/api.py 
CONFLICT (content): Merge conflict in torch/distributed/rpc/api.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 1]
```diff
@@ -575,6 +575,7 @@ def test_wait_all_workers(self):
         # worker1/2 calls this immediately and has some works after it.
         # worker3 calls this immediately and has no more work.
         rpc.api._wait_all_workers()
+        rpc.api._wait_all_workers()
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: let's split this into two tests, call it once and call it twice.

File: torch/distributed/rpc/api.py [hunk 2]
```diff
@@ -76,39 +77,56 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+class WaitAllWorkersStates(object):
+    def __init__(self):
+        # Each `intent_worker_names` is an empty set at beginning.
+        # It's only used by leader worker. Leader worker is user-specified or
+        # elected as the first worker in a sorted worker name list.
+        # Whenever there is a worker showing shutdown intention to the leader, by
+        # calling _wait_all_workers()`, the leader adds this worker's name to the set.
```
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: missing a "`" at the before _wait_all_workers()

**Review by mrshenli — APPROVED** [review]
LGTM!

**Pull request closed without merging**